### PR TITLE
Implemented microphone modes

### PR DESCRIPTION
**Pull request by LexiconCode**
## Description

Implemented microphone modes to toggle microphone on, off, and sleep.

The new commands
`caster <mode>` on, sleep, off (DNS/DPI only)

These modes use grammar/rule exclusivity to block other grammars from being recognized. 
- Kaldi uses still recognize dictation but no action will be executed. The same is true for DNS/DPI but it's not printed out the window.
- DNS/DPI is also a bit special if you change the mic mode from the DNS/DPI GUI it will stay in sync with Caster mic modes
- DNS/DPI built-in commands for waking up and going to sleep are still functional as they rely on exclusivity internally. There is no method that I know of to disable these internal commands. 

These modes are implemented through 2 rules. The 1st rule `CasterMicRule` to execute the function to enable exclusivity. The 2nd rule `sleep_rule` leverages native dragonfly to become exclusive via `ExclusiveManager` overriding `CasterMicRule` and all other rules loaded into the engine. `sleep_rule` uses dragonfly `FuncContext` to toggle on and off. Both rules use `EngineModesManager` to manage the different modes based on the engine is running. There are some significant differences. DNS utilizes Natlink and `_sync_mode` to synchronize modes then uses ExclusiveManager. The other engines utilize just ExclusiveManager. 

**Note** because native dragonfly was used for `sleep_rule` simplified transformers cannot be used on `CasterMicRule` or `sleep_rule`.

Could this implementation be simplified?
 If we could create separate CCR merger that's for exclusive grammars only it would drastically simplify the implementation.

What about additional modes?
These are some thoughts for the future but not in this PR. I did not want to take longer to get the sleep mode out to users as it impacts the usability of Caster.
- Implement set_engine_mode exclusivity with mode rules. dictation, numbers, and spell for all engines
- Implement timer for sleep mode for all engines but DNS/DPI.
- Implement hotkey for microphone on-off

## Related Issue

https://github.com/dictation-toolbox/Caster/issues/797

## Motivation and Context

Motivation for this is to unify microphone modes command regardless of the speech recognition backend for users.

## How Has This Been Tested

I've tested this with DNS and Kaldi switching between microphone modes.

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue or bug)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Renamed existing command phrases (we discourage this without a strong rationale).

## Checklist

- [x] I have read the CONTRIBUTING document.
- [x] My code follows the code style of this project.
- [x] I have checked that my code does not duplicate functionality elsewhere in Caster.
- [x] I have checked for and utilized existing command phrases from within Caster (delete if not applicable). 
- [x] My code implements all the features I wish to merge in this pull request.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests pass.

## Maintainer/Reviewer Checklist

- [x] Basic functionality has been tested and works as claimed.
- [ ] New documentation is clear and complete.
- [ ] Code is clear and readable.

### CODE REVIEW

File: castervoice/rules/core/engine_manager_rules/mic_rules.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+from castervoice.lib.ctrl.mgr.engine_manager import EngineModesManager
+
+class CasterMicRule(MappingRule):
+    mapping = {
+        "caster <mic_mode>": Function(lambda mic_mode: EngineModesManager.set_mic_mode(mode=mic_mode)),     
+    }
+    extras = [
+        Choice(
+            "mic_mode", {
+                "off": "off",
```
Comment (kendonB) [inline castervoice/rules/core/engine_manager_rules/mic_rules.py:16]: I get this when saying "caster off"

`Caster: 'off' is not valid. set_mic_state modes are: 'off' - DPI Only, 'on', 'sleeping'`

Is that deliberate? We should also have "caster wake" as an option.

Comment (kendonB) [reply castervoice/rules/core/engine_manager_rules/mic_rules.py:16]: @LexiconCode thanks for doing this!

Comment (LexiconCode) [reply castervoice/rules/core/engine_manager_rules/mic_rules.py:16]: Yes it is intentional ('off' - DPI Only). For other engines besides DPI/DNS to have an off state they need a method to turn the microphone on without voice activation like a keyboard shortcut/GUI toggle. DPI/DNS has that capability.

**Comment by esc123** [comment]
Using DPI 15.3 on Windows 10, all three commands (on, off, sleep) work perfectly. Good stuff.

**Comment by kendonB** [comment]
@LexiconCode is there an easy way to implement automatic sleeping functionality? I often forget that the system is on after using it for a few minutes and end up making it do a bunch of crazy things because I'm speaking around it or listening to something.

**Comment by LexiconCode** [comment]
> @LexiconCode is there an easy way to implement automatic sleeping functionality? I often forget that the system is on after using it for a few minutes and end up making it do a bunch of crazy things because I'm speaking around it or listening to something.

Yes I think that shouldn't be too bad to do. It would involve creating a timer that is reset every time there's a successful recognition.

**Pull request merged**